cmd/lotus-provider/run.go

```diff
@@ -0,0 +1,284 @@
+				_, err := lr.Datastore(context.Background(), "/metadata")
+				if err != nil {
+					return err
+				}
+			}
+			if err := lr.Close(); err != nil {
+				return xerrors.Errorf("close repo: %w", err)
+			}
+		}
+
+		lr, err := r.Lock(repo.Provider)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := lr.Close(); err != nil {
+				log.Error("closing repo", err)
+			}
+		}()
+
+		db, err := harmonydb.NewFromConfig(config.HarmonyDB{
+			Username: cctx.String("db_user"),
+			Password: cctx.String("db_password"),
+			Hosts:    strings.Split(cctx.String("db_host"), ","),
+			Database: cctx.String("db_name"),
+			Port:     cctx.String("db_port"),
+		})
+		if err != nil {
+			return err
+		}
+		// TODO add harmonytask
+		_ = db
+
+		shutdownChan := make(chan struct{})
+
+		stop, err := node.New(ctx,
+			node.Override(new(dtypes.ShutdownChan), shutdownChan),
+			node.Provider(r),
+		)
+		if err != nil {
+			return xerrors.Errorf("creating node: %w", err)
+		}
+
+		const unspecifiedAddress = "0.0.0.0"
+		address := cctx.String("listen")
+		addressSlice := strings.Split(address, ":")
+		if ip := net.ParseIP(addressSlice[0]); ip != nil {
+			if ip.String() == unspecifiedAddress {
+				timeout, err := time.ParseDuration(cctx.String("timeout"))
+				if err != nil {
+					return err
+				}
+				rip, err := extractRoutableIP(timeout)
+				if err != nil {
+					return err
+				}
+				address = rip + ":" + addressSlice[1]
+			}
+		}
+		localStore, err := paths.NewLocal(ctx, lr, nil, []string{"http://" + address + "/remote"})
+		if err != nil {
+			return err
+		}
+
+		handler := mux.NewRouter()
+		fh := &paths.FetchHandler{Local: localStore, PfHandler: &paths.DefaultPartialFileHandler{}}
+		handler.NotFoundHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !auth.HasPerm(r.Context(), nil, api.PermAdmin) {
+				w.WriteHeader(401)
+				_ = json.NewEncoder(w).Encode(struct{ Error string }{"unauthorized: missing admin permission"})
+				return
+			}
+
+			fh.ServeHTTP(w, r)
+		})
+		// local APIs
+		{
+			m := mux.NewRouter()
+			// debugging
+			m.Handle("/debug/metrics", metrics.Exporter())
+			m.PathPrefix("/").Handler(http.DefaultServeMux) // pprof
+
+			var hnd http.Handler = m
+
+			handler.PathPrefix("/").Handler(hnd)
+		}
+
+		// Serve the RPC.
+		endpoint, err := r.APIEndpoint()
+		if err != nil {
+			return xerrors.Errorf("getting API endpoint: %w", err)
+		}
+		rpcStopper, err := node.ServeRPC(handler, "lotus-provider", endpoint)
+		if err != nil {
+			return fmt.Errorf("failed to start json-rpc endpoint: %s", err)
+		}
+
+		// Monitor for shutdown.
+		finishCh := node.MonitorShutdown(shutdownChan,
+			node.ShutdownHandler{Component: "rpc server", StopFunc: rpcStopper},
+			node.ShutdownHandler{Component: "provider", StopFunc: stop},
+		)
+
+		<-finishCh
+		return nil
+	},
+}
+
+func extractRoutableIP(timeout time.Duration) (string, error) {
+	conn, err := net.DialTimeout("udp", "8.8.8.8:80", timeout)
```

Comment:
This won't work in firewalled environments or in places on earth where 8.8.8.8 isn't easily reachable.

We will need a connection to some chain node rpc from lotus-provider in all(?) cases, so it should be reasonable to repurpose the method from lotus-worker to use the chain node connection.

(I also saw a helper to get this from the DB connection, that also seems reasonable)

Reply:
The DB helper is the way to go